### PR TITLE
[RFC] Establish TVM Unity Connection -- A Technical Strategy

### PR DESCRIPTION
**Pull request by tqchen**
This RFC summarizes an overall technical strategy on establish TVM Unity connection milestone.
It is supplementary to the technical RFCs about specific components.

[rendered](https://github.com/tqchen/tvm-rfcs/blob/main/rfcs/0091-establish-tvm-unity-connection.md)
[discuss thread](https://discuss.tvm.apache.org/t/establish-tvm-unity-connection-a-technical-strategy/13344)

### CODE REVIEW

**Comment by Hzfengsy** [comment]
Thanks @tqchen!!! I'm excited to see the pre-RFC become this formal RFC.  

The Unity Connection is a great step from multi-level lowering compilation to a flexible, unified abstraction for the end-to-end model compilation. I'd like to summarize the [discuss thread](https://discuss.tvm.apache.org/t/establish-tvm-unity-connection-a-technical-strategy/13344) here for readers who did not participate.

## Modularized Compilation Flow
The TVM unity uses cross-layer abstraction to represent:
1. Graph IR: how to organize ops/kernels
2. Tensor IR/Libraries/FFI: how to execute the ops/kernels
Based on such abstraction, we can build the module at any stage as long as the module is legal. However, the current multi-stage lowering pipeline requires we must have a `GraphIR->TensorIR->RuntimeModule` pipeline.

## Easy to customize 
Fast customization is critical during researching and prototyping. The modularized workflow natively enables it. Here I'd like to share two cases:

### Ex1: Adding new operator supports
Instead of [7-step tutorial](https://tvm.apache.org/docs/dev/how_to/relay_add_op.html), here are only two steps with the unity connection:
- implement how the op is computed (both tir or libraries are good),
- Directly call the implementation in the unified abstraction.

### Ex2: Customizable operator fusion
Each pass is decoupled, which means we are able to fuse operators (to be general, optimize the module) in multiple passes. i.e. We can have a customized pass to fuse two convs while using the internal fusor to fuse the following element-wise ops.

## Cross-layer optimization opportunities
Layout optimization is a typical cross-layer optimization, that we are able to do with TVM Unity. Also, we do have some prototype results to prove it works. Additionally, I'm happy to see the community members working on different backends are all looking forward to this feature.

Note that this RFC is a technical strategy, which is a bit different with the Relax Upstream RFC https://github.com/apache/tvm-rfcs/pull/89. Please turn to that thread if you have specific comments on relax itself.

Love to hear ideas from the community.

File: rfcs/0091-establish-tvm-unity-connection.md [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+To help stage the set of improvements while minimizing disruption of the current flow. We propose and introduce unity connection as an optional compilation flow of TVM, diagrammed as:
+
+![Untitled 1|690x276](assets/0091/unity_connection_diagram.png) 
+
+In the proposal, the “default build” (those that run through `tvmc` and `relay.build`) that TVM users know today will remain unchanged. TVM unity connection will be added as an optional vertical component. 
+
+This is a change that is invisible to the current users who use the existing flow as the default build will continue to work as-is. In the meantime, we give our community members (many of whom are excited about TVM unity) abilities to leverage unity connection early as an optional vertical to enable use cases that may not be supported today (such as dynamic shapes, direct topi integration, and cross abstraction optimizations).
```
Comment (zhiics) [inline rfcs/0091-establish-tvm-unity-connection.md:133]: I think it is not just about the community members, but also the users and developers who are not in the community but interested in checking it out.

File: rfcs/0091-establish-tvm-unity-connection.md [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+There are many advantages of unifying BYOC as IRModule⇒IRModule pass:
+
+- We can choose when to run BYOC, and interpolate multiple BYOC runs if necessary.
+- Smart search and auto-tuning can be built on top as separate passes that make the decision and dispatch without overburdening the infrastructure.
+- The overall connection point is the IRModule spec: we represent BYOC result as call into PackedFunc and IRModules in `external_mods` attrs, we can explore many ways of smart BYOC planning without increasing the complexity of the interface.
+
+### Framework Fallbacks and Library Integration
+
+Framework integration is another important thing that gets simplified through the unity connection. Right now we face challenges to support a broad set of operators that the framework support. One approach is to quickly fall back to framework libraries themselves. In other cases, we face the need to leverage in-house libraries, such as cuDNN.  Previously, the choices of the libraries are hidden behind the dispatching strategies in various places and tied to a single build pipeline. Adding support for a new fallback operator involves changing multiple places.
```
Comment (zhiics) [inline rfcs/0091-establish-tvm-unity-connection.md:279]:
```suggestion
Framework integration is another important thing that gets simplified through the unity connection. Right now we face challenges to support a broad set of operators that are supported by frameworks. One approach is to quickly fall back to framework libraries themselves. In other cases, we face the need to leverage in-house libraries, such as cuDNN.  Previously, the choices of the libraries are hidden behind the dispatching strategies in various places and tied to a single build pipeline. Adding support for a new fallback operator involves changing multiple places.
```

File: rfcs/0091-establish-tvm-unity-connection.md [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+```python
+@tvm.script.ir_module
+class MyModWithLibTorch:							
+    @R.func
+    def main(x: R.Tensor((1, 10, 32, 32))):
+        w0 = R.const(shape=(10, 32, 3, 3))
+        lv0 = R.call_dps_packed(R.extern("torch.conv2d"), [x, w], (1, 20, 32, 32))
+```
+
+`torch.conv2d` is registered as a PackedFunc in the runtime environment and leverages DLPack to enable the exchange of data between TVM runtime and PyTorch in a zero-copy fashion. We can also go further to automate the generation of such environment bindings, supporting most of the library functions that torch support and automatically fall back when the operator is not available.
```
Comment (zhiics) [inline rfcs/0091-establish-tvm-unity-connection.md:292]:
```suggestion
`torch.conv2d` is registered as a PackedFunc in the runtime environment and leverages DLPack to enable the exchange of data between TVM runtime and PyTorch in a zero-copy fashion. We can also go further to automate the generation of such environment bindings, supporting most of the library functions that torch supports and automatically fall back when the operator is not available.
```

File: rfcs/0091-establish-tvm-unity-connection.md [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+        call_packed(matmul, x, w0, lv1)
+        call_packed(matmul, lv1, w1, out)
+        ...
+```
+
+The above code is an example. By solidifying the IRModule spec, we can simplify our development by making assumptions explicit in the form of IR, and build connections (transformations) from the IRModule that contains graph-TIR to this form.
+
+### Customize Fusion
+
+As we start to experiment with more advanced optimizations, there is usually a need to customize various behaviors in the pipeline. Fusion is one such example. Right now the fusion passes support fusing elementwise operations onto complicated operations such as conv2d. Let us imagine that we want to customize fusion to support fusing two conv2d together. It was a non-trivial task to hack the fusion engines and moves the needles around them.
```
Comment (zhiics) [inline rfcs/0091-establish-tvm-unity-connection.md:326]:
```suggestion
As we start to experiment with more advanced optimizations, there is usually a need to customize various behaviors in the pipeline. Fusion is one such example. Right now the fusion passes support fusing elementwise operations onto complicated operations such as conv2d. Let us imagine that we want to customize fusion to support fusing two conv2d together. It was a non-trivial task to hack the fusion engines and move the needles around them.
```

File: rfcs/0091-establish-tvm-unity-connection.md [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+### Customized Operator Support and Direct TOPI Integration
+
+As we start to bring in more ML models, there is a need to enable a broad spectrum of models. Ease of support for customized operators is a common need. Unity offers `irbuilder.emit_te(topi.matmul, inputs)`, which effectively takes a TOPI function and constructs a call into the relevant TIR function. The direct TOPI integration enables us to quickly build up custom op through TVMScript/TE without surgical changes to the codebase.
+
+### Evolving Default Build 
+
+This RFC proposes a first step for us to establish Unity Connection as an optional end-to-end compilation flow of TVM. The TVM unity connection keeps the default build flows of `tvmc` and `relay.build` unchanged in this RFC. This section describes possible approaches to evolve the default build to leverage more of unity components. Note again that this is only a discussion of possible approach is not part of the RFC, but only to bring possible relevant context of future opportunties.
+
+The exact approach to take and detailed discussions shall happen in dedicated RFCs for each incremental change to the default build.
```
Comment (zhiics) [inline rfcs/0091-establish-tvm-unity-connection.md:376]:
```suggestion
The detailed discussions about the exact approach shall happen in dedicated RFCs for each incremental change to the default build.
```

File: rfcs/0091-establish-tvm-unity-connection.md [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+One approach to enable unity connection is to start from the low-level part of the default build. Enabling a greater amount of models going from relay (which optimizes graph op) to relax. We will then bring more capabilities to the unity connection while simplifying some of the low-level components.
```
Comment (zhiics) [inline rfcs/0091-establish-tvm-unity-connection.md:378]:
```suggestion
One approach to enable unity connection starts from the low-level part of the default build. Enabling a greater amount of models going from Relay (which optimizes graph op) to Relax. We will then bring more capabilities to the unity connection while simplifying some of the low-level components.
```

**Review by zhiics — COMMENTED** [review]
I've already read the pre-RFC in the discussion forum and shared some thoughts there. Here I only have some comments on the grammatical nits and typos.

**Review by ZihengJiang — APPROVED** [review]
Looks great to me and I don't have further opinion since I have already left my comments in the discussion forum. Excited to see this happens and looking forward to those key components being landed.

**Comment by tqchen** [comment]
opened a voting thread https://github.com/apache/tvm/issues/12651

**Comment by tqchen** [comment]
Thanks, everyone, for putting effort into making unity development happen. 

Today, we come to the one-year mark of the unity connection proposal. It is amazing to see how the landscape of AI/ML has changed and how some of the emerging needs fit into the strategy we bought one year ago. Because of the time delays. The strategy as it is no longer timely sense as it was one year ago. 

So we decided to withdraw the proposal

Standing at last year's point, it is obvious that if we took the default option of not pursuing or delaying the strategy, we would miss the opportunity to empower the community in the age of genAI. 

Luckily, we did not let that slip away and pushed the development of  [unity branch](https://discuss.tvm.apache.org/t/establish-tvm-unity-branch/14244). Many of you may have seen the news lately on how they bear fruits in enabling real applications like bringing LLMs and other foundational models to various devices.

The good news is that we have clarified [strategy decision process](https://github.com/apache/tvm-rfcs/blob/main/rfcs/0102-clarify-strategy-decision-process.md) to empower strategic proposals. There are also new emerging opportunities in the age of genAI and foundational models, which [majority community support](https://discuss.tvm.apache.org/t/discuss-tvm-community-strategy-for-foundational-models/15401).

Please follow the unity branch, continue participating in community discussions, and let us bring TVM to enable everyone. Thank you to many community members who supported the proposal and believed in what it can bring. Let us do our best pushing unity development. We will also collect inputs and come up with new versions of strategies that reflects the current inputs.

**Comment by junrushao** [comment]
Thanks everyone for the year-long discussion.

I'd love to note that, in respective, we would have already missed the boat of generative AI, and Apache TVM would have lost momentum of empowering the community on the up-to-date workloads they are interested in, if we decided to follow the default option not pursuing a real solution, and to stagnate with endless debates without concrete action taken.

I'm super glad to have a set of decisive PMC staff who care and a broad vibrant community who contribute and help. We work collaborative on Unity, this super valuable technical directions, and catch up and empower generative AI together!

**Pull request closed without merging**